### PR TITLE
Remove Zend\Stdlib dependency

### PR DESCRIPTION
**Pull request by snapshotpl**
...and simplify config implementation

### CODE REVIEW

**Comment by Ocramius** [comment]
@snapshotpl to also be removed from `composer.json`

**Comment by snapshotpl** [comment]
@Ocramius that funny because Stdlib doesn't exists in composer.json on master :)

**Comment by snapshotpl** [comment]
It's very strange for me why current tests on master pass. When I pull master branch, run `composer update` and run test I get:
```
PHP Fatal error:  Class 'Zend\Stdlib\ArrayUtils' not found in zend-servicemanager/src/Config.php on line 62
```
How it's possible that travis said it's ok? Is it cache?

File: src/Config.php [hunk 1]
```diff
@@ -46,27 +27,20 @@ class Config implements ConfigInterface
...
      * @param array $config
      */
     public function __construct(array $config = [])
     {
         // Only merge keys we're interested in
-        foreach (array_keys($config) as $key) {
-            if (! isset($this->allowedKeys[$key])) {
-                unset($config[$key]);
+        foreach (array_keys($this->config) as $requiredKey) {
+            if (isset($config[$requiredKey])) {
```
Comment (weierophinney) [inline src/Config.php]: This should likely be a combination of `isset && is_array`, as every single item is expected to be an array. I'll make that change on merge.

File: src/Config.php [hunk 2]
```diff
@@ -9,28 +9,9 @@
...
-use Zend\Stdlib\ArrayUtils;
-
 class Config implements ConfigInterface
 {
     /**
-     * Allowed configuration keys
-     *
-     * @var array
-     */
-    protected $allowedKeys = [
```
Comment (Ocramius) [inline src/Config.php]: This is a BC break tho :-(

Dammit, we need to stop this `protected` madness.

**Comment by weierophinney** [comment]
Oof, on review, I realized that this is a rather large BC break, and cannot be merged. Why? Because it changes the behavior.

In the current stable version, if values are present in the `$config` property already, then any values provided via the passed `$config` array to the constructor are *merged* with those values. With this patch, they *replace* them, which is not and never was the intention.

The appropriate way to get rid of zend-stdlib is as @Ocramius suggested elsewhere: copy the functionality of `ArrayUtils::merge` into a trait or directly into the `Config` implementation. If you would like to amend this patch or provide a new one to do so, I'll review that.

**Comment by snapshotpl** [comment]
Sounds bad. Implementation of `ArrayUrils::merge` is more complicated than just copy - it uses other classes/interfaces (eg https://github.com/zendframework/zend-stdlib/blob/master/src/ArrayUtils.php#L272). I think we should add stdlib to dependency, or make BC break as it's enought early to nobody start use v3 :see_no_evil: :hear_no_evil: :speak_no_evil: 
What are you think? Same with `$allowedKeys`

**Comment by snapshotpl** [comment]
@weierophinney BTW I think is bug - tests on master don't pass.

**Comment by Ocramius** [comment]
Only arrayutils::merge needs polyfilling, btw.

**Comment by mtymek** [comment]
@weierophinney @Ocramius what about extracting ArrayUtils as a separate micro-library, same as it was done with hydrators? This way we will avoid code duplication. Yes, one external dependency would be still there, but we won't load lots of unnecessary classes.

**Comment by Maks3w** [comment]
I think StdLib is enough small and don't need to be splitted.

**Comment by Ocramius** [comment]
Agreed, stdlib is small enough

**Comment by mtymek** [comment]
It is small if you take number of classes, but it has many responsibilities:

* ArrayUtils
* StringUtils
* interfaces for request and response
* priority list, priority queue
* ...

**Comment by Ocramius** [comment]
Wondering if we can just c&p the merging code (as private api)

**Comment by snapshotpl** [comment]
Ok, but what about this feature? https://github.com/zendframework/zend-stdlib/blob/master/src/ArrayUtils.php#L272 Without this it's still break

**Comment by bakura10** [comment]
Can't it be simplified based on the context? For instance for a lib contains only a specific use case just Optimize the arrayUtils by removing useless stuff?

Envoyé de mon iPhone

**Comment by snapshotpl** [comment]
Ok, I copy ArrayUtils::merge to Config. I worried about `MergeReplaceKeyInterface` and `MergeRemoveKey`, however class doesn't need to exists to chech instance, so it's working. I have write unit test and I have found some issue in my implementation. I fixed that, however current implementation works same and looks simpler so I bring it back

**Comment by weierophinney** [comment]
I was concerned about the `MergeReplaceKeyInterface` and `MergeRemoveKey` usage, but realized that these are:

- advanced features
- opt-in features

We can make a suggestion in the `composer.json` indicating that if those features are needed, the user must install zend-stdlib.

Will merge shortly!

**Comment by Ocramius** [comment]
Note that the breakage still needs documenting, but it is indeed part of
protected API, so not a very huge deal

Marco Pivetta

http://twitter.com/Ocramius

http://ocramius.github.com/

**Comment by weierophinney** [comment]
@Ocramius Documenting it in the CHANGELOG. I've also documented the features in the class-level docblock of the `Config` class, and added a `suggest` entry; you'll see this in the merge commit shortly.

**Pull request merged**